test/integration/apiserver/certreload/aggregator_certreload_test.go

```diff
@@ -0,0 +1,341 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package certreload
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	genericapiserveroptions "k8s.io/apiserver/pkg/server/options"
+	client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/cert"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	aggregatorclient "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+	kastesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	apiserverintegration "k8s.io/kubernetes/test/integration/apiserver"
+	"k8s.io/kubernetes/test/integration/framework"
+	testutil "k8s.io/kubernetes/test/utils"
+	sampleserver "k8s.io/sample-apiserver/pkg/cmd/server"
+)
+
+func TestAggregatorCertReload(t *testing.T) {
+	proxyCA, proxyClientKey, proxyClientCert, err := newCAWithClientCert()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var proxyClientKeyPath string
+	var proxyClientCertPath string
+	var servingCertPath string
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	t.Log("STEP 1: Start Kube API Server")
+	testServer := kastesting.StartTestServerOrDie(t, &kastesting.TestServerInstanceOptions{EnableCertAuth: true, ModifyServerRunOptions: func(opts *options.ServerRunOptions) {
+		servingCertPath = opts.SecureServing.ServerCert.CertDirectory
+
+		proxyClientKeyPath = filepath.Join(servingCertPath, "proxy-client.key")
+		if err := ioutil.WriteFile(proxyClientKeyPath, proxyClientKey, 0644); err != nil {
+			t.Fatal(err)
+		}
+
+		proxyClientCertPath = filepath.Join(servingCertPath, "proxy-client.pem")
+		if err := ioutil.WriteFile(proxyClientCertPath, testutil.EncodeCertPEM(proxyClientCert), 0644); err != nil {
+			t.Fatal(err)
+		}
+
+		opts.ProxyClientCertFile = proxyClientCertPath
+		opts.ProxyClientKeyFile = proxyClientKeyPath
+		opts.EnableAggregatorRouting = true
```

Comment:
what of these tweaks cannot be done with flags?

Reply:
Removing instead we can use flags to apiserver:

--proxy-client-key-file string
--proxy-client-cert-file string
--enable-aggregator-routing

Reply:
@sttts I have rewritten the code by using flags but now trying to figure out how to pass the cert-file and key-file files to the testServer when the file paths are by default created inside StartTestServer and we generate cert as part of ModifyServerRunOptions.

Also you have suggested to remove: ModifyServerRunOptions, how do we pass `dynamiccertificates.FileRefreshDuration = 1 * time.Second `as a kube-apiserver argument?

Thank you for your help, I ll also check with @p0lyn0mial  and see what would be the best to do.

Reply:
We usually pre-create certs in tests because generation costs entropy and might block or slow down tests. Does that help with the directory problem?

Reply:
We definitely don't want a generic and powerful modifier like ModifyServerRunOptions. This will encourage misuse. If there is something super internal that is not exposed via flags (like maybe fresh duration), make it an explicit field  of TestServerInstanceOptions.

Reply:
> We usually pre-create certs in tests because generation costs entropy and might block or slow down tests. Does that help with the directory problem?

Thank you! I was not aware, meaning I can use key and crt in  testdata folder: https://github.com/kubernetes/kubernetes/tree/master/cmd/kube-apiserver/app/testing/testdata

Reply:
@sttts I have pre-generated the certs, do you think it makes sense to expose testing.pkgPath? I want to use that one to locate testdata, rather than using my own.

Reply:
Please review new changes in this commit https://github.com/kubernetes/kubernetes/pull/93021/commits/d30d3903741bf0832d0bf6ca10def65f15b5dcf9

I have pre-generated the certs and removed the ModifyServerRunOptions